F/ist

p4u opened the pull request:
Compute results as consensus would allow for the computation of results based on consensus among the nodes on the Vochain blockchain. This would ensure that the results are accurate and tamper-proof, and would also prevent any single node from having too much influence over the outcome. Currently, the results are not part of the state, but by making the Vochain nodes compute and verify the results, the results would be much more verifiable.

To enable this feature, we must introduce a new internal state transition mechanism (IST), which would make the Vochain transition to a new state without the need of external transactions. However all nodes execute the IST at exactly same height. Thus we need to create a new component, part of the state named Internal State Transition Controller (ISTC).

p4u committed:
introduce the IST controller

ISTC is the internal state transition controller.

It is a new component of the Vochain designed to
perform internal transition states.

These transitions are scheduled for a specific height.

The first use case is the internal results computation.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
remove oracle references (1/2)

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
Merge remote-tracking branch 'origin/master' into f/ist

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
upgrade protobuf

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
remove oracle code (2/2) and fix on isSynchronizing

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
update scrutinizer to use IST results

Signed-off-by: p4u <pau@dabax.net>

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+
+// Controller is the internal state transition controller.
+type Controller struct {
+	st              *state.State
```
use a longer name, otherwise `m.st` is not very useful :) perhaps `m.state`?

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+
+// Action is the model used to store the IST actions into state.
+type Action struct {
+	Action     int
```
assuming this gets marshalled, I would say pick either int32 or int64, otherwise the behavior may be different depending on the architecture.

mvdan replied:
this field name also stutters; perhaps something like `Action.ID` or `Action.Number` instead?

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+// Actions is the model used to store the list of IST actions for
+// a specific height into state.
+type Actions struct {
+	Actions map[string]Action
```
any reason to make this a field of a struct, instead of `type Actions map[string]Action`?

or if you think you might need more fields in the future, rename the field, like `type Actions struct { ByName map...` to avoid the `Actions.Actions` stutter.

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+// encode performs the encoding of the IST action using Gob.
+func (m *Actions) encode() []byte {
+	var buf bytes.Buffer
+	if err := gob.NewEncoder(&buf).Encode(m); err != nil {
```
remember that encoding/gob is not a stable format; it may change with future Go versions. I think it's fine for in-memory stuff or temporary caches, but anything relatively permanent should use a stable format. For that, I'd either use go-bare if we don't need a schema, or protobuf if we want a schema.

p4u replied:
Since this information is not relevant for the Consensus State (its stored into a NoState() KV space), it is not a big issue if the Gob encoder in some point changes to a different format.

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+	// find the IST action
+	var actID string
+	for actID = range actions.Actions {
+		if string(id) == actID {
```
why are you looping over a map to find a key? just index into the map?

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+
+// Reschedule reschedules an IST action to the given new height.
+func (c *Controller) Reschedule(id []byte, oldHeight, newHeight uint32) error {
+	if id == nil {
```
probably do `if len(id) == 0` instead, otherwise you'll happily accept `Reschedule(make([]byte, 0), ...` which is still an invalid ID

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+	// anything else, smooth return.
+	if process.Status == models.ProcessStatus_CANCELED ||
+		process.Status == models.ProcessStatus_RESULTS ||
+		process.Status == models.ProcessStatus_ENDED {
```
use a switch :)

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+	// ActionCommitResults schedules the commit of the results to the state.
+	ActionCommitResults
+	// ActionEndProcess sets a process as ended. It schedules ActionComputeResults.
+	ActionEndProcess
```
make these typed constants, to ensure that you don't mix them with other integer values

mvdan commented on vochain/ist/store.go:
```diff
@@ -0,0 +1,52 @@
+func dbIndex(height uint32) []byte {
+	hBytes := make([]byte, 4)
+	binary.LittleEndian.PutUint32(hBytes, height)
+	return append([]byte(dbPrefix), hBytes...)
```
Or, since Go 1.19, you can now do: `return binary.LittleEndian.AppendUint32([]byte(dbPrefix), height)` :)

mvdan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,303 @@
+			if !isSynchronizing {
+				// compute results in a goroutine to avoid blocking the state transition
+				go func() {
+					if err := c.computeAndStoreResults(action.ElectionID); err != nil {
```
you might have multiple of these calls happen concurrently, would that be safe?

p4u replied:
Yes, it should be safe.

mvdan reviewed (commented):
would be nice to also document which of these new `ist` methods are OK to call concurrently; APIs by default are not safe for concurrent use unless the docs say so.

p4u committed:
Merge remote-tracking branch 'origin/master' into f/ist

p4u committed:
apply PR suggestions

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
add some hysteresis to the synchronizing call

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
remove oracle from readme

Signed-off-by: p4u <pau@dabax.net>

mariajdab commented on vochain/ist/results.go:
```diff
@@ -0,0 +1,113 @@
+	}
+
+	// Delete the stored results
+	_ = c.deleteFromNoState(dbResultsIndex(electionID))
```
Just curious about why is not necessary check a possible error? Or maybe a log could be useful

p4u replied:
The reason is, there might be a situation where there are no results to delete (if the first condition of the loop is met). And in any case, if the deletion of results fails for some real reason, probably the database is corrupted and more errors will raise.

mariajdab replied:
Got it! Thanks

mariajdab commented on vochain/ist/results.go:
```diff
@@ -0,0 +1,113 @@
+	return nil
+}
+
+// commitResults commits the results to the state. If the reults parameter is nil, it will
```
typo `reults`

p4u committed:
Merge remote-tracking branch 'origin/master' into f/ist

Signed-off-by: p4u <pau@dabax.net>

mariajdab commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,298 @@
+func (m *Actions) encode() []byte {
+	var buf bytes.Buffer
+	if err := gob.NewEncoder(&buf).Encode(m); err != nil {
+		// should never happend
```
should never happen :eyes:

p4u committed:
upgrade docker compose test suit

- move prometheus/grafana to the main compose
- fix grafana/prometheus and add a default dashboard
- use a network between nodes instead of link (deprecated)
- minor tweaks

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
fix some typos

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
add back the new tests

Signed-off-by: p4u <pau@dabax.net>

jordipainan commented on cmd/end2endtest/main.go:
```diff
@@ -103,6 +103,8 @@ func main() {
 			fmt.Fprintf(os.Stderr, op.example+"\n")
 			fmt.Fprintf(os.Stderr, "\n")
 		}
+		fmt.Fprintf(os.Stderr, "If the network is deployed locally using the docker test suit, ")
```
```suggestion
		fmt.Fprintf(os.Stderr, "If the network is deployed locally using the docker test suite, ")
```

jordipainan commented on cmd/node/main.go:
```diff
@@ -92,7 +91,7 @@ func newConfig() (*config.Config, config.Error) {
 		"overwrite an existing config file with the provided CLI flags")
 	globalCfg.Mode = *flag.StringP("mode", "m", types.ModeGateway,
-		"global operation mode. Available options: [gateway,oracle,ethApiOracle,miner,seed]")
+		"global operation mode. Available options: [gateway,miner,seed]")
```
```suggestion
		"global operation mode. Available options: [gateway, miner, seed]")
```

jordipainan commented on config/config.go:
```diff
@@ -5,7 +5,7 @@ import (
 )
 
-// DvoteCfg stores global configs for dvote
+// Config stores global configs for dvote
```
```suggestion
// Config stores global configurations for vocdoni-node
```

jordipainan commented on db/batch.go:
```diff
@@ -43,7 +43,7 @@ func NewBatchMaxSize(db Database, maxSize uint) *Batch {
 }
 
-// Get implements the WriteTx.Get interface method
+// Get implements the WriteTx.Get interface method.
```
```suggestion
// Get implements the WriteTx.Get() interface method.
```

jordipainan commented on dockerfiles/testsuite/env.miner0:
```diff
@@ -9,4 +9,4 @@ VOCDONI_VOCHAIN_GENESIS=/app/misc/genesis.json
 VOCDONI_METRICS_ENABLED=True
 VOCDONI_METRICS_REFRESHINTERVAL=5
-VOCDONI_VOCHAIN_MINERTARGETBLOCKTIMESECONDS=5
+VOCDONI_VOCHAIN_MINERTARGETBLOCKTIMESECONDS=6
```
Why this change ? Is it because you want to keep the refresh interval lower that the target block time ?

p4u replied:
Because 6 is the reality. I have not been able to achieve 5s

mvdan commented on apiclient/helpers.go:
```diff
@@ -191,17 +191,17 @@ func (c *HTTPclient) WaitUntilTxIsMined(ctx context.Context,
 		case <-ctx.Done():
 			return nil, fmt.Errorf("election %s keys not yet published: %w", electionID, ctx.Err())
+		default:
+			time.Sleep(PollInterval)
```
why this change? it seems like the old code was much better. if the context is cancelled, the sleep should stop - that's what time.After did.

p4u replied:
tbh, I don't now. Probably this is some merge with master that went wrong.

p4u replied:
~~Ah, I think this is part of @mariajdab PR (that was merged recently)~~
No, this was me....

jordipainan commented on dockerfiles/testsuite/env.miner1:
```diff
@@ -1,12 +1,12 @@
 VOCDONI_METRICS_ENABLED=True
 VOCDONI_METRICS_REFRESHINTERVAL=5
-VOCDONI_VOCHAIN_MINERTARGETBLOCKTIMESECONDS=5
+VOCDONI_VOCHAIN_MINERTARGETBLOCKTIMESECONDS=6
```
Same

jordipainan commented on dockerfiles/testsuite/env.miner2:
```diff
@@ -1,11 +1,11 @@
 VOCDONI_METRICS_ENABLED=True
 VOCDONI_METRICS_REFRESHINTERVAL=5
-VOCDONI_VOCHAIN_MINERTARGETBLOCKTIMESECONDS=5
+VOCDONI_VOCHAIN_MINERTARGETBLOCKTIMESECONDS=6
```
Same

jordipainan commented on dockerfiles/testsuite/env.miner3:
```diff
@@ -1,11 +1,11 @@
 VOCDONI_METRICS_ENABLED=True
 VOCDONI_METRICS_REFRESHINTERVAL=5
-VOCDONI_VOCHAIN_MINERTARGETBLOCKTIMESECONDS=5
+VOCDONI_VOCHAIN_MINERTARGETBLOCKTIMESECONDS=6
```
Same

mvdan commented on dockerfiles/testsuite/env.miner2:
```diff
@@ -1,11 +1,11 @@
 VOCDONI_DATADIR=/app/run
 VOCDONI_MODE=miner
-VOCDONI_LOGLEVEL=info
+VOCDONI_LOGLEVEL=debug
```
wouldn't this make CI way too chatty?

mvdan commented on vochain/indexer/migrations/0001_create_table_processes.sql:
```diff
@@ -11,7 +11,6 @@ CREATE TABLE processes (
   results_votes           TEXT NOT NULL,
   results_weight          TEXT NOT NULL,
   results_envelope_height INTEGER NOT NULL,
-  results_signatures      TEXT NOT NULL,
```
this is a hard breaking change on the indexer DB - all existing DBs will need to be deleted. are we still OK with that?

p4u replied:
Yes. This PR is breaking everything anyway. So we need a hard reset.

lucasmenendez commented on dockerfiles/testsuite/genesis.json:
```diff
@@ -4,7 +4,7 @@
   "consensus_params": {
     "block": {
-      "max_bytes": "5242880",
+      "max_bytes": "2097152",
```
Just a doubt, what does this limit mean and why is it being reduced?

p4u replied:
This is the size of the block. So the number of transactions that can be included in a period of time of 10-12 seconds.

I've performed several tests and benchmarking before creating this PR and 5 MiB seems too much. The block time easily increases to 30-40 seconds on high load (thousands of votes per minute continuously). However if using 2 MiB, the network performs stable, with a block time of 12-13 seconds on high load.

lucasmenendez commented on vochain/app.go:
```diff
@@ -42,13 +42,14 @@ var (
 	// tendermint WaitSync() function is racy, we need to use a mutex in order to avoid
 	// data races when querying about the sync status of the blockchain.
-	isSyncLock sync.Mutex
+	isSynchronizing uint32
```
Why not just use `atomic.Uint32`?

p4u replied:
No reason. But its kind of the same, no?

lucasmenendez replied:
Yes, I think they are both the same, but probably using the atomic version would be clearer. That way, just by reading the variable declaration you know that it is atomic, currently you have to look into the `IsSynchronizing()` method to know that.

mvdan replied:
you could also simply export an https://pkg.go.dev/sync/atomic#Bool, removing the need for uint32, the Fn field, your own method, etc.

lucasmenendez reviewed (approved):
Great job! The improvements make the process shorter and easier to understand, at least for me.

p4u committed:
review fixes

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
Update cmd/end2endtest/main.go

Co-authored-by: Jordi Pinyana  <jordi.pinyana.paga@gmail.com>

p4u committed:
Update config/config.go

Co-authored-by: Jordi Pinyana  <jordi.pinyana.paga@gmail.com>

p4u committed:
Update db/batch.go

Co-authored-by: Jordi Pinyana  <jordi.pinyana.paga@gmail.com>

p4u committed:
Update cmd/node/main.go

Co-authored-by: Jordi Pinyana  <jordi.pinyana.paga@gmail.com>

jordipainan commented on dockerfiles/testsuite/start_test.sh:
```diff
@@ -121,6 +117,8 @@ e2etest_cenususizelection () {
 log "### Starting test suite ###"
 $COMPOSE_CMD build
+$COMPOSE_CMD up -d seed # start the seed first so the nodes can properly bootstrap
+sleep 10
```
10 sec seems too much, I think 5 sec is actually enough

p4u replied:
Its not. The seed needs to be able to bootstrap the network, in a slow environment such as some CIs, it might require around 10 seconds.

jordipainan commented on statedb/treeupdate.go:
```diff
@@ -129,6 +130,12 @@ func (u *TreeUpdate) Set(key, value []byte) error {
 }
 
+// Del removes a key-value from this tree.  `key` is the path of the leaf.
+func (u *TreeUpdate) Del(key []byte) error {
```
I assume every time you delete a key the tree "rebalances". Do you think that so many deletes can impact on tx processing performance?

p4u replied:
The Del method is only used for the NoState() database currently.

jordipainan commented on statedb/treeview.go:
```diff
@@ -30,6 +30,12 @@ func (v *databaseViewer) Get(key []byte) ([]byte, error) {
 }
 
+// IterateWithPrefix iterates over all the keys that start with the given prefix.
+// The iration stops when the callback returns false.
```
```suggestion
// The iteration stops when the callback returns false.
```

jordipainan commented on tree/tree.go:
```diff
@@ -121,6 +121,24 @@ func (t *Tree) Add(wTx db.WriteTx, key, value []byte) error {
 	return nil
 }
 
+// Del deletes a key from the Tree.
```
What about: 

```
// Del deletes a key from the Tree.
func (t *Tree) Del(wTx db.WriteTx, key []byte) error {
	givenTx := wTx == nil
	if givenTx {
		wTx = t.DB().WriteTx()
		defer wTx.Discard()
	}
	if err := wTx.Delete(key); err != nil {
		return err
	}
	if givenTx {
		return wTx.Commit()
	}
	return nil
}
```
- If the negation operator is removed the code is more readable (IMO)
- In the last condition check we can just return the `wtx.Commit()` so there is no need to check the error

jordipainan commented on vochain/genesis/genesis.go:
```diff
@@ -118,8 +104,8 @@ var devGenesis = GenesisDoc{
 var stageGenesis = GenesisDoc{
-	GenesisTime: time.Date(2023, time.March, 10, 10, 0, 0, 0, time.UTC),
-	ChainID:     "vocdoni-stage-3",
+	GenesisTime: time.Date(2023, time.April, 10, 10, 0, 0, 0, time.UTC),
```
Please next time make it clear that DEV & STG envs will be affected as these are breaking consensus changes.

p4u replied:
All commits will be squashed. The final message will contain the breaking notice.

jordipainan commented on vochain/indexer/indexer_test.go:
```diff
@@ -530,10 +531,7 @@ func TestResults(t *testing.T) {
-		Nonce: fmt.Sprintf("%x", util.RandomBytes(32)),
-		Votes: []int{1, 1, 1, 1},
-	})
+	vp, err := state.NewVotePackage([]int{1, 1, 1, 1}).Encode()
```
Oh! that is a cool one! what a pleasure to create a vote pkg like this <3

jordipainan commented on vochain/indexer/indexer_test.go:
```diff
@@ -1251,10 +1238,7 @@ func TestOverwriteVotes(t *testing.T) {
-		Nonce: fmt.Sprintf("%x", util.RandomHex(32)),
-		Votes: []int{0, 0, 0},
-	})
+	vp, err = state.NewVotePackage([]int{0, 0, 0}).Encode()
```
What happens if I overflow the int value ?

p4u replied:
In Go, the numeric overflows do not panic, instead it starts over from the zero value (so it's like applying modulus over the max possible value).

jordipainan commented on vochain/indexer/vote.go:
```diff
@@ -214,50 +210,42 @@ func (s *Indexer) GetEnvelopeHeight(processID []byte) (uint64, error) {
 
-// ComputeResult process a finished voting, compute the results and saves it in the Storage.
+// finalizeResults process a finished voting, get the results from the state and saves it in the indexer Storage.
 // Once this function is called, any future live vote event for the processId will be discarted.
```
```suggestion
// Once this function is called, any future live vote event for the processId will be discarded.
```

jordipainan commented on vochain/indexer/vote.go:
```diff
@@ -214,50 +210,42 @@ func (s *Indexer) GetEnvelopeHeight(processID []byte) (uint64, error) {
+func (s *Indexer) finalizeResults(processID []byte) error {
 	height := s.App.Height()
-	log.Debugf("computing results on height %d for %x", height, processID)
+	log.Debugw("finalize results", "processID", hex.EncodeToString(processID), "height", height)
```
I think the concept of finalize results should be better explained.

p4u replied:
What you mean? This is a debug message. Its scope is to provide technical information to the developer.

jordipainan replied:
Sorry I was not referring to this debug message in particular, but added the message here. I was referring to the concept of finalizing the results. It is a new concept at least for me and was not easy to understand the change at first glance. That is why I suggested to add more context, maybe on top of the function.

jordipainan commented on vochain/state/process.go:
```diff
@@ -237,8 +237,8 @@ func (v *State) SetProcessStatus(pid []byte, newstatus models.ProcessStatus, com
 	case models.ProcessStatus_ENDED:
-		if currentStatus != models.ProcessStatus_READY {
-			return fmt.Errorf("process %x can only be ended from ready status", pid)
+		if currentStatus != models.ProcessStatus_READY && currentStatus != models.ProcessStatus_PAUSED {
```
Why this?

p4u replied:
I think it makes more sense. I can see a use case where the organizer pauses the process, waits some time to have some kind of confirmation, and then ENDs it directly. Actually, I don't see a reason to not support this transition. 
Do you have any strong point against it?

jordipainan replied:
Oki, nope any strong opinion. Just wanted to understand the context. Thank you :)

p4u committed:
Update statedb/treeview.go

Co-authored-by: Jordi Pinyana  <jordi.pinyana.paga@gmail.com>

p4u committed:
Update vochain/indexer/vote.go

Co-authored-by: Jordi Pinyana  <jordi.pinyana.paga@gmail.com>

p4u committed:
apply PR suggestions

Signed-off-by: p4u <pau@dabax.net>

jordipainan commented on vochain/ist/ist.go:
```diff
@@ -0,0 +1,298 @@
+			// check if we have all the revealed keys, else reschedule
+			if !c.checkRevealedKeys(action.ElectionID) {
+				// we are missing some keys, we cannot compute the results yet
+				newHeight := height + (1 + action.Attempts*2)
```
Maybe it is useful creating a function for this. Something like `RescheduleHeight()`

jordipainan commented on vochain/results/compute.go:
```diff
@@ -0,0 +1,120 @@
+	rawVote := make([]byte, len(VotePackage))
+	copy(rawVote, VotePackage)
+	// if encryption keys, decrypt the vote
+	if len(keys) > 0 {
```
I do think that this condition is not required, the loop wont be executed if len(keys) = 0

jordipainan commented on vochain/results/compute.go:
```diff
@@ -0,0 +1,120 @@
+
+	var nvotes atomic.Uint64
+	lock := sync.Mutex{}
+	starTime := time.Now()
```
typo `startTime`

jordipainan commented on vochain/results/compute.go:
```diff
@@ -0,0 +1,120 @@
+	}
+	results.EnvelopeHeight = nvotes.Load()
+	return results, err
+}
```
```suggestion
	startTime := time.Now()
	err = st.IterateVotes(electionID, true, func(vote *models.StateDBVote) bool {
		keys := []string{}
		if p.EnvelopeType.EncryptedVotes {
			for _, k := range vote.EncryptionKeyIndexes {
				if k < types.KeyKeeperMaxKeyIndex && k < len(p.EncryptionPrivateKeys) {
					keys = append(keys, p.EncryptionPrivateKeys[k])
				} else {
					log.Warn("wrong vote: key index overflow or too many fields")
					return false
				}
			}
		}
		vp, err := unmarshalVote(vote.VotePackage, keys)
		if err != nil {
			log.Debugf("vote invalid: %v", err)
			return false
		}
		if err = results.AddVote(vp.Votes, new(big.Int).SetBytes(vote.Weight), &lock); err != nil {
			log.Warnf("addVote failed: %v", err)
			return false
		}
		nvotes.Add(1)
		return false
	})
	if err == nil {
		log.Infow("computed results",
			"process", fmt.Sprintf("%x", electionID),
			"votes", nvotes.Load(),
			"results", results.String(),
			"elapsed", time.Since(startTime).String(),
		)
	}
	results.EnvelopeHeight = nvotes.Load()
	return results, err
}
```

- Solved typo `starTime`
- You can remove unnecessary else block and initialize keys as an empty slice before the condition check. Thus reducing the nesting level
- You can remove the separate conditions for "encryptionKeyIndexes has too many fields" and "key index overflow", and merge them into a single condition.

jordipainan reviewed (approved):
I'm so happy that we can finally remove the Oracle. Thank you so much for the hard effort you putted here.
It is a very big task.
I think the IST logic is correct and I did not find any edge case after the review.
Some functions in IST related code can be optimized but nothing is a big deal, I made some suggestions but feel free to ignore them.

Thank you <3

p4u committed:
Update vochain/results/compute.go

Co-authored-by: Jordi Pinyana  <jordi.pinyana.paga@gmail.com>
Signed-off-by: p4u <pau@dabax.net>
